### ISSUE TITLE
Lowering Sails causes error connecting to DB

### ISSUE THREAD

**Issue by crobinson42** [balderdashy/sails-mongo#398]
When I lower sails this error persists:

```
TypeError: Cannot read property 'collections' of undefined
    at Object.module.exports.adapter.find (/therms-1337/node_modules/sails-mongo/lib/adapter.js:348:40)
    at [object Object].module.exports.find (/therms-1337/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:119:13)
    at [object Object]._runOperation (/therms-1337/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:411:29)
    at [object Object].run (/therms-1337/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:70:8)
    at [object Object].module.exports.find (/therms-1337/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/basic.js:264:16)
    at [object Object].wrapper [as find] (/therms-1337/node_modules/lodash/index.js:3095:19)
    at [object Object].Deferred.exec (/therms-1337/node_modules/sails/node_modules/waterline/lib/waterline/query/deferred.js:485:16)
    at Object.socketDisconnected (OnlineController.js:80:5)
    at Object.wrapper [as socketDisconnected] (/therms-1337/node_modules/lodash/index.js:3095:19)
    at Object.socketDisconnected (SocketService.js:157:32)
    at Object.wrapper [as socketDisconnected] (/therms-1337/node_modules/lodash/index.js:3095:19)
    at SocketService.js:117:14
    at /therms-1337/node_modules/sails/lib/hooks/session/index.js:30:16
    at Command.callback (/therms-1337/node_modules/connect-redis/lib/connect-redis.js:136:25)
    at normal_reply (/therms-1337/node_modules/connect-redis/node_modules/redis/index.js:662:21)
    at RedisClient.return_reply (/therms-1337/node_modules/connect-redis/node_modules/redis/index.js:722:9)
    at JavascriptReplyParser.RedisClient.reply_parser.Parser.returnReply (/therms-1337/node_modules/connect-redis/node_modules/redis/index.js:146:18)
    at JavascriptReplyParser.run (/therms-1337/node_modules/connect-redis/node_modules/redis/node_modules/redis-parser/lib/javascript.js:137:18)
    at JavascriptReplyParser.execute (/therms-1337/node_modules/connect-redis/node_modules/redis/node_modules/redis-parser/lib/javascript.js:112:10)
    at Socket.<anonymous> (/therms-1337/node_modules/connect-redis/node_modules/redis/index.js:223:27)
    at emitOne (events.js:82:20)
    at Socket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at Socket.Readable.push (_stream_readable.js:110:10)
    at TCP.onread (net.js:529:20)
```

Is there any recommendation on a shim I can place during my build process or direction for contributing a PR for this fix?

**Does anyone else experience this?**

**Comment by sailsbot** [comment]
@crobinson42 Thanks for posting, we'll take a look as soon as possible.  In the meantime, if you haven’t already, please carefully read the [issue contribution guidelines](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md#ii-issue-contributions) and double-check for any missing information above.  In particular, please ensure that this issue is about a stability or performance bug with a  documented feature; and make sure you’ve included detailed instructions on how to reproduce the bug from a clean install.   Finally, don’t forget to include the version of Node.js you tested with, as well as your version of Sails or Waterline, and of any relevant standalone adapters/generators/hooks.

Thank you!

**Comment by crobinson42** [comment]
Sails v0.12.1
Sails-mongo v0.12.0
Node v4.4.2
Npm v2.14

This happens on local machine connecting to Mongo DB with recommended configuration and also happens on Linux production box connecting to a remote Mongo DB connection.

**Comment by crobinson42** [comment]
```
^[[A/Users/coryrobinson/projects/therms/node_modules/sails-mongo/lib/adapter.js:348
      var collection = connectionObject.collections[collectionName];
                                       ^

TypeError: Cannot read property 'collections' of undefined
    at Object.module.exports.adapter.find (/Users/coryrobinson/projects/therms/node_modules/sails-mongo/lib/adapter.js:348:40)
    at [object Object].module.exports.find (/Users/coryrobinson/projects/therms/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:119:13)
    at [object Object]._runOperation (/Users/coryrobinson/projects/therms/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:411:29)
    at [object Object].run (/Users/coryrobinson/projects/therms/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:70:8)
    at [object Object].module.exports.find (/Users/coryrobinson/projects/therms/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/basic.js:264:16)
    at [object Object].wrapper [as find] (/Users/coryrobinson/projects/therms/node_modules/lodash/index.js:3095:19)
    at [object Object].Deferred.exec (/Users/coryrobinson/projects/therms/node_modules/sails/node_modules/waterline/lib/waterline/query/deferred.js:485:16)
    at Object.socketDisconnected (OnlineController.js:80:5)
    at Object.wrapper [as socketDisconnected] (/Users/coryrobinson/projects/therms/node_modules/lodash/index.js:3095:19)
    at Object.socketDisconnected (SocketService.js:157:32)
    at Object.wrapper [as socketDisconnected] (/Users/coryrobinson/projects/therms/node_modules/lodash/index.js:3095:19)
    at SocketService.js:117:14
    at /Users/coryrobinson/projects/therms/node_modules/sails/lib/hooks/session/index.js:30:16
    at Command.callback (/Users/coryrobinson/projects/therms/node_modules/connect-redis/lib/connect-redis.js:136:25)
    at normal_reply (/Users/coryrobinson/projects/therms/node_modules/connect-redis/node_modules/redis/index.js:662:21)
    at RedisClient.return_reply (/Users/coryrobinson/projects/therms/node_modules/connect-redis/node_modules/redis/index.js:722:9)
```

**Comment by no7dw** [comment]
same happened here:
npm : v2.1.17
node : v0.10.35
"sails": "0.12.1"
"sails-mongo": "0.12.0",

    TypeError: Cannot read property 'collections' of undefined
        at Object.module.exports.adapter.create (/var/git/finance/fin_sails_bear/node_modules/sails-mongo/lib/adapter.js:300:40)
        at module.exports.create (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:84:13)
        at createValues (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/query/dql/create.js:220:16)
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/query/dql/create.js:74:20
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:726:13
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:52:16
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:269:32
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:44:16
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:723:17
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:167:37
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:52:16
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:269:32
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:44:16
        at null.<anonymous> (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/utils/schema.js:152:44)
        at fn (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/utils/callbacksRunner.js:41:10)
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:181:20
        at iterate (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:262:13)
        at Object.async.forEachOfSeries.async.eachOfSeries (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:281:9)
        at Object.async.forEachSeries.async.eachSeries (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:214:22)
        at Object.runner.beforeCreate (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/utils/callbacksRunner.js:44:9)
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/query/dql/create.js:180:17
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:718:13


also appear sth like here:

    TypeError: Cannot call method 'find' of undefined
        at Object.module.exports.adapter.find (/var/git/finance/fin_sails_bear/node_modules/sails-mongo/lib/adapter.js:351:18)
        at module.exports.find (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:119:13)
        at _runOperation (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:411:29)
        at run (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:70:8)
        at module.exports.find (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/basic.js:264:16)
        at wrapper [as find] (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/lodash/index.js:3095:19)
        at Object.module.exports.find (/var/git/finance/fin_sails_bear/api/controllers/App_versionController.js:21:17)
        at wrapper (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/lodash/index.js:3095:19)
        at routeTargetFnWrapper (/var/git/finance/fin_sails_bear/node_modules/sails/lib/router/bind.js:179:5)
        at callbacks (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:164:37)
        at param (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:138:11)
        at param (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:135:11)
        at pass (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:145:5)
        at nextRoute (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:100:7)
        at callbacks (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:167:11)
        at /var/git/finance/fin_sails_bear/node_modules/sails/lib/router/bind.js:187:7
        at alwaysAllow (/var/git/finance/fin_sails_bear/node_modules/sails/lib/hooks/policies/index.js:158:11)
        at routeTargetFnWrapper (/var/git/finance/fin_sails_bear/node_modules/sails/lib/router/bind.js:179:5)
        at callbacks (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:164:37)
        at param (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:138:11)
        at param (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:135:11)
        at pass (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/express/lib/router/index.js:145:5)

some relative:

    Error: Error: Connection is already registered
        at /var/git/finance/fin_sails_bear/node_modules/sails/lib/hooks/orm/index.js:242:19
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/async/lib/async.js:52:16
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/async/lib/async.js:565:21
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/async/lib/async.js:167:37
        at /var/git/finance/fin_sails_bear/node_modules/sails/lib/hooks/orm/build-orm.js:53:23
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline.js:206:21
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:52:16
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:570:21
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:167:37
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:52:16
        at done (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:241:17)
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:44:16
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline.js:199:27
        at Object.module.exports.adapter.registerConnection (/var/git/finance/fin_sails_bear/node_modules/sails-redis/lib/adapter.js:65:51)
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline.js:198:29
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:181:20
        at Object.async.forEachOf.async.eachOf (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:233:13)
        at Object.async.forEach.async.each (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:209:22)
        at Array.async.auto.registerConnections [as 1] (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/lib/waterline.js:171:13)
        at /var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:596:38
        at _arrayEach (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:85:13)
        at Object.async.auto (/var/git/finance/fin_sails_bear/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:554:9)


recently we upgrade sails version and sails-mongo version , these error didn't happen before.

**Comment by no7dw** [comment]
we found out that this issue is probably relative with sails-hook-autoreload
Once the code is change ,it happens . During these 3 days ever since we removed in prod  it hasn't happened again.

**Comment by crobinson42** [comment]
> we found out that this issue is probably relative with sails-hook-autoreload
I've tried disabling and also commenting out `sails-hook-autoreload` and the issue persists.

**Comment by crobinson42** [comment]
Ok, this issue, for me at least, is related to sockets afterDisconnect() which writes to my DB when a client disconnects. If I have  a browser open and shut the server down, I'll see this error.

**Comment by sailsbot** [comment]
@crobinson42,@sailsbot,@no7dw: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).



If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!
